Refactored LimitedAmmo to AmmoPool

reaperrr opened the pull request:
...and moved `Reloads` functionality and `RearmSound` to `AmmoPool`.

Spiritual successor to #6514, but changed approach quite a bit in order to avoid most of the issues #6514 had and slightly improve functionality on top of that.

reaperrr committed:
Refactors LimitedAmmo to AmmoPool.

Removes Reloads trait.

This enables adding multiple AmmoPools via @ differentiators and
Name which adds the possibility to assign each armament to a specific
ammo pool.

Furthermore, this moves all Reloads functionality onto AmmoPool.

Now a combination of all three is possible on a single actor: no limited
ammo, limited ammo that can reload on its own, and limited ammo which
needs to be reloaded at a rearm actor.

Additionally moves RearmSound from Minelayer to AmmoPool.

reaperrr committed:
Upgrade rules for LimitedAmmo -> AmmoPool refactoring

penev92 commented on OpenRA.Mods.Common/Activities/Air/HeliAttack.cs:
```diff
@@ -24,10 +24,19 @@ public override Activity Tick(Actor self)
-			var reloads = self.TraitOrDefault<Reloads>();
-			if (limitedAmmo != null && !limitedAmmo.HasAmmo() && reloads == null)
-				return Util.SequenceActivities(new HeliReturn(), NextActivity);
+			var limitedAmmo = self.TraitsImplementing<LimitedAmmo>();
```
My three comments on `FlyAttack` also apply ti this file.

penev92 replied:
My three comments on `FlyAttack` also apply to this file.

penev92 commented on OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:
```diff
@@ -28,9 +28,18 @@ public override Activity Tick(Actor self)
-			var limitedAmmo = self.TraitOrDefault<LimitedAmmo>();
-			if (limitedAmmo != null && !limitedAmmo.HasAmmo())
-				return NextActivity;
+			var limitedAmmo = self.TraitsImplementing<LimitedAmmo>();
```
I'd rather add `Actor self` to the constructor and cache this there so we don't run it on every Tick. Maybe also rename it to `limitedAmmoTraits`?

reaperrr replied:
> I'd rather add Actor self to the constructor and cache this there so we don't run it on every Tick.

OK. If I'm not mistaken, it might make sense to do the same for ```attack``` for the same reason, right?

penev92 replied:
Yes, I think you're right.

penev92 commented on OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:
```diff
@@ -28,9 +28,18 @@ public override Activity Tick(Actor self)
+			var limitedAmmo = self.TraitsImplementing<LimitedAmmo>();
+			if (limitedAmmo != null)
+			{
+				foreach (var la in self.TraitsImplementing<LimitedAmmo>())
```
You already got these above ^.
Don't repeat that operation.

penev92 commented on OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:
```diff
@@ -28,9 +28,18 @@ public override Activity Tick(Actor self)
+
+					return NextActivity;
+				}
+			}
```
I believe we don't want this block, since it's redundant.

reaperrr replied:
Um, I'm not sure what you're referring to.

penev92 replied:
The brackets in the `if`, around the `foreach` - you can do without them.

reaperrr replied:
Oh, I see.
If I remember correctly (which isn't guaranteed, given my poor memory), we use brackets when there's more than 1 line following the ```if```, for the sake of readability.

penev92 replied:
Some people argue that every `if` or loop statement should have brackets. In OpenRA we use them only when we have to. And if you have more than one line, you **need** to group the lines using brackets. (the `foreach` here and it's block are considered one line)

reaperrr replied:
Alright, I'll remove them.

penev92 commented on OpenRA.Mods.Common/Activities/Rearm.cs:
```diff
@@ -17,44 +17,60 @@ namespace OpenRA.Mods.Common.Activities
-			remainingTicks = ticksPerPip;
-			this.sound = sound;
+			{
+				foreach (var la in self.TraitsImplementing<LimitedAmmo>())
```
Same as above comments.

penev92 commented on OpenRA.Mods.Common/Activities/Rearm.cs:
```diff
@@ -17,44 +17,60 @@ namespace OpenRA.Mods.Common.Activities
+				{
+					ticksPerPip = la.ReloadTimePerAmmo();
+					remainingTicks = ticksPerPip;
+					this.sound = sound;
```
These three will have the values of the last trait in `limitedAmmo` after the loop is over. You might be missing something here?   :?

reaperrr replied:
Well, that's not the intended result, so yes, I'm missing something here. I don't know what exactly I need to change to fix this, though.

penev92 replied:
I'm thinking you will need to either run a separate `Rearm` activity for each of `self.TraitsImplementing<LimitedAmmo>()`, or have a global `limitedAmmoTraits` collection in `Rearm` and populate it in this loop, then iterate through it in `Tick()`.

reaperrr replied:
Sorry for having to ask again, but how do I do this (the second one) or where can I get an example?

penev92 replied:
I'd be happy to help, but this isn't really a very convenient place for that. Come to IRC and look for me :D

penev92 commented on OpenRA.Mods.Common/Activities/Rearm.cs:
```diff
@@ -17,44 +17,60 @@ namespace OpenRA.Mods.Common.Activities
 
 		public override Activity Tick(Actor self)
 		{
+			var limitedAmmo = self.TraitsImplementing<LimitedAmmo>();
```
Same as above comments - cache this in the constructor.

penev92 commented on OpenRA.Mods.Common/Traits/Armament.cs:
```diff
@@ -46,10 +49,8 @@ public class ArmamentInfo : UpgradableTraitInfo, ITraitInfo, Requires<AttackBase
 
 		[Desc("Muzzle flash sequence to render")]
 		public readonly string MuzzleSequence = null;
-
```
New lines are nice. Please separate all these fields with new lines while you are at it.

Phrohdoh replied:
Don't mix coding style fixes/clarity (if they don't change logic) with logical changes.
If you are going to do this create a new commit.

penev92 replied:
Usually I'd agree, but we're talking about 5 new lines in a class he already changes, 2 of which are being removed by him. Making another PR for those would probably cause the code to be left with no new lines (pretty much how it was left the last time someone changed these files).
Plus it seems redundant.

cjshmyr replied:
> Don't mix coding style fixes/clarity (if they don't change logic) with logical changes. If you are going to do this create a new commit.

I disagree, this isn't a confusing or complicated change.

Phrohdoh replied:
@penev92 I didn't say PR I said commit.
@cjshmyr I'm just relaying what I've been told many times by @chrisforbes.

penev92 replied:
Ah OK.

penev92 commented on OpenRA.Mods.Common/Traits/LimitedAmmo.cs:
```diff
@@ -17,58 +17,106 @@ namespace OpenRA.Mods.Common.Traits
+		[Desc("Time to reload automatically measured in ticks.")]
+		public readonly int SelfReloadTicks = 25 * 2;
+		[Desc("Whether or not reload counter should be reset when ammo has been fired.")]
+		public readonly bool ResetOnFire = false;
```
Again - add tons of new lines, please.

penev92 commented on OpenRA.Mods.Common/Traits/LimitedAmmo.cs:
```diff
@@ -17,58 +17,106 @@ namespace OpenRA.Mods.Common.Traits
+		[Sync] public int CurrentAmmo;
+		[Sync] public int RemainingTicks;
+		public int PreviousAmmo;
 		LimitedAmmoInfo info;
```
The `info` field should probably be readonly.

penev92 commented on OpenRA.Mods.RA/AI/HackyAI.cs:
```diff
@@ -310,7 +310,7 @@ public bool HasMinimumProc()
 		// For mods like RA (number of building must match the number of aircraft)
 		bool HasAdequateAirUnits(ActorInfo actorInfo)
```
We should rename this to 'HasEnoughAirfields', or if you prefer - 'HasEnoughAircraftReloadingBuildings' :D

penev92 commented on OpenRA.Mods.RA/AI/HackyAI.cs:
```diff
@@ -310,7 +310,7 @@ public bool HasMinimumProc()
 		bool HasAdequateAirUnits(ActorInfo actorInfo)
 		{
-			if (!actorInfo.Traits.Contains<ReloadsInfo>() && actorInfo.Traits.Contains<LimitedAmmoInfo>()
+			if (actorInfo.Traits.Contains<LimitedAmmoInfo>() 
```
This method used to check if the actor needs an airfield to reload and if there are enough airfields for all actors of this type.
You need to add a check for `!ReloadsAutomatically()' to stay true to the idea of the method.

reaperrr replied:
What I've been struggling with is that this is calling for ActorInfo rather than Actor.

Would ```!actorInfo.Traits.Get<LimitedAmmoInfo>().SelfReloads``` work?
It does compile, at least...

penev92 commented:
Left several comments on the code. The last one should cover be what you needed help with (the AI check).  ;)

reaperrr commented:
Updated. Rearm has been refactored courtesy of @penev92.
I wasn't sure whether I should keep the move of RearmSound for a follow-up PR, but it's a relatively simple change and implementing a hacky stop-gap work-around for the minelayer would have been at least as much work, so I stuffed it in here.

penev92 commented:
>Now a combination of all three is possible on a single actor: no limited
ammo, limited ammo that can reload on its own, and limited ammo which
needs to be reloaded at a rearm actor.

I think you should add an example for all 3 to a mod of your choice so we can test/review this easier.
Btw I tested yesterday with the Longbow you sent me and it worked great. ;)

reaperrr commented:
> I think you should add a few (the more - the better) combination examples to a mod of your choice so we can test/review this easier. These will of course be removed before merging this.

Ok, but that will have to wait until tomorrow.

penev92 commented on OpenRA.Mods.Common/Activities/Rearm.cs:
```diff
@@ -8,53 +8,78 @@
+		readonly IEnumerable<LimitedAmmo> limitedAmmoPools;
 
-		public Rearm(Actor self, string sound = null)
+		// Keep a collection of all the pools with a number representing the remaining reload time for each
```
This is probably redundant.

penev92 commented on OpenRA.Mods.Common/Activities/Rearm.cs:
```diff
@@ -8,53 +8,78 @@
+			// TRUE if we have nothing more to reload
+			var needsReloading = false;
+
+			// iterate through all pools
```
This is redundant.

penev92 commented on OpenRA.Mods.Common/Activities/Rearm.cs:
```diff
@@ -8,53 +8,78 @@
+
+				needsReloading = true;
+
+				// if it is time to reload ... do something
```
This is redundant.

penev92 commented on OpenRA.Mods.Common/Activities/Rearm.cs:
```diff
@@ -8,53 +8,78 @@
+				// if it is time to reload ... do something
+				if (--limitedAmmoPoolsReloadTimes[pool] == 0)
+				{
+					// BOY is this a hack... well... HACK to check if we are on the helipad/airfield/etc.
```
Change that to just
`// BOY is this a hack...`
:D

penev92 commented on OpenRA.Mods.Common/Activities/Rearm.cs:
```diff
@@ -8,53 +8,78 @@
+					// BOY is this a hack... well... HACK to check if we are on the helipad/airfield/etc.
+					var hostBuilding = self.World.ActorMap.GetUnitsAt(self.Location).FirstOrDefault(a => a.HasTrait<RenderBuilding>());
+
+					// If we are not... move on
```
Redundant comment.

penev92 commented on OpenRA.Mods.Common/Activities/Rearm.cs:
```diff
@@ -8,53 +8,78 @@
 
-				if (hostBuilding == null || !hostBuilding.IsInWorld)
-					return NextActivity;
+					// Give me 1 more "ammo" if the pool is not full
```
Redundant comment.

penev92 commented:
I think nobody will want those comments I left, so better remove them. Also I'll do a more proper review once I can test this.

penev92 commented on OpenRA.Mods.Common/Traits/LimitedAmmo.cs:
```diff
@@ -17,58 +17,121 @@ namespace OpenRA.Mods.Common.Traits
+		[Desc("How much ammo is reloaded after a certain period.")]
+		public readonly int ReloadCount = 0;
+
+		[Desc("How much ammo is reloaded after a certain period.")]
```
Fix description.

reaperrr commented:
@penev92: I'll remove the comments and fix the description later.
For now, I've added testcases to RA mod (Mammoth and Hind), and more importantly, added a WIP fix for Repairable, we forgot/overlooked that. Doesn't work as intended yet, I just wanted to get rid of the crash for now.

penev92 commented:
By the way that `LimitedAmmo` dependency that `Repairable` has is really weird and should not exist IMO.

Phrohdoh commented on OpenRA.Mods.Common/Traits/LimitedAmmo.cs:
```diff
@@ -17,58 +17,121 @@ namespace OpenRA.Mods.Common.Traits
+		[Desc("Whether or not ammo is replenished on its own.")]
+		public readonly bool SelfReloads = false;
+
+		[Desc("Time to reload automatically measured in ticks.")]
```
Either remove "measured in ticks" or rename this variable, please.
Duplication is unnecessary.

Phrohdoh commented on OpenRA.Mods.Common/Traits/LimitedAmmo.cs:
```diff
@@ -17,58 +17,121 @@ namespace OpenRA.Mods.Common.Traits
-		public bool HasAmmo() { return ammo > 0; }
+		public string AmmoPoolName { get { return info.AmmoPoolName; } }
+		public string RearmSound { get { return info.RearmSound; } }
+		public bool ReloadsAutomatically() { return info.SelfReloads == true; }
```
`== true` is unnecessary.

All of these "getters" for the *Info properties are unnecessary bloat, just make the Info public.

Mailaender commented:
Also needs a rebase.

penev92 commented:
@reaperrr I see you are still around (glad to see that :) )
What are your plans regarding this PR?
It's close to being ready, so I'd hate to have to close it for future generations.

reaperrr commented:
@penev92 I just had a lot of stress at work recently and caught a cold on top of that, draining me of the 'head-bandwidth' necessary for dealing with this. I'll update & rebase once things settle down a little.

reaperrr commented:
Updated and rebased.

Still some open issues (Repair, possibly checks in AirStates, and maybe squashing some commits).

Phrohdoh commented:
```
OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:L28: [SP2000] Invalid spacing at the end of the line.
OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:L29: [SP2000] Invalid spacing at the end of the line.
OpenRA.Mods.Common/Activities/Air/HeliAttack.cs:L27: [SP2000] Invalid spacing at the end of the line.
OpenRA.Mods.Common/Activities/Air/HeliAttack.cs:L28: [SP2000] Invalid spacing at the end of the line.
OpenRA.Mods.Common/Traits/LimitedAmmoPool.cs:L95: [SP2000] Invalid spacing at the end of the line.
OpenRA.Mods.Common/Traits/LimitedAmmoPool.cs:L96: [SP2000] Invalid spacing at the end of the line.
OpenRA.Mods.Common/Traits/LimitedAmmoPool.cs:L98: [SP2000] Invalid spacing at the end of the line.
OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:L711: [SP2000] Invalid spacing at the end of the line.
```

reaperrr commented:
Updated.

penev92 commented:
What is the "Help wanted" label for?

reaperrr commented:
Updated, fixed some stupid snafus from yesterday.

> What is the "Help wanted" label for?

Nevermind, last time I tested this there seemed to be a problem with ordering the testcase mammoth to rearm on the service depot, but when I tested it a moment ago it seemed to work.

This should be ready for a final round of reviewing & testing, then.

obrakmann commented:
Got this while playtesting:

```
Exception of type `System.InvalidOperationException`: TypeDictionary contains multiple instance of type `OpenRA.Mods.Common.Traits.LimitedAmmoPoolInfo`
  at OpenRA.Primitives.TypeDictionary.Get[LimitedAmmoPoolInfo] () [0x00089] in /home/oliver/devel/openra/git/OpenRA.Game/Primitives/TypeDictionary.cs:63 
  at OpenRA.Mods.Common.AI.HackyAI.HasAdequateAirUnitReloadBuildings (OpenRA.ActorInfo actorInfo) [0x00023] in /home/oliver/devel/openra/git/OpenRA.Mods.Common/AI/HackyAI.cs:313 
```

reaperrr commented:
Updated.

obrakmann commented:
Aside from the mentioned style nits, this looks good, and works ingame, :+1:

reaperrr commented:
Updated, adressed style nits.

> It would be nice if this could only skip to the next activity if the ammo for the weapon used to run the attack is empty. But I guess that will make this a lot more complicated, so this is just wishful thinking. Feel free to ignore.

In my opinion there should actually be an option for the modder to choose between "returns when all ammo usable against target is depleted" and "returns once the first ammo pool usable against the target is depleted", but I'd prefer to deal with that in a future PR, otherwise this would get delayed even more than it already has.

The current solution is sufficient for the shipping mods, so it should be okay for now.

penev92 commented:
Also I do apologize that all those comments above were on the commit and not on the PR :(

reaperrr commented:
> Yeah, sorry that comment was so vague. @Phrohdoh explained it. And while I suppose the minelayer will never have a second ammo pool, it mught be nice to check for consistency with other traits.
On the other hand the minelayer has its own trait which noone else uses...
But I still vote for consistency with the rest.

I've already dealt with this locally (will push this with some other changes tonight), and minelayers still work so the code can't be too bad :)
It even checks against an AmmoPool property in ```Minelayer``` now.

I'm dropping a summary of open issues here so they won't get lost when I update the commits later:
```Rearm.cs```
- L31: possibly turn into single Select()
- L52: check building name against actor's rearming structure

```AmmoPool.cs```
- L97: invert if (might be ok the way it is, but needs to be checked)
- L110: should go before L97 according to @penev92 (see above)
- L120+: possibly simplify ```GetPips```

```Repair.cs```
- discuss & implement best approach (for this PR)

Phrohdoh commented:
@reaperrr are you waiting for input or review?
@penev92 what changes are you requesting?

It doesn't make sense to have `Needs +2` and `Changes requested`.

penev92 commented:
Wow all my comments are gone :X
Guess I'll have to do another review. @reaperrr was waiting for help (which I promised, but haven't gotten around to yet :/ ) with fixing my requested changes.

reaperrr commented:
> Edit: Oh, @reaperrr saved the comments :)

Was about to write that ;) 
I knew they'd get lost otherwise.

Phrohdoh commented:
He saved the gist of them but nothing had happened, which is why I am prodding. :)

penev92 commented:
Working on it now ;)

penev92 commented on OpenRA.Mods.Common/Traits/AmmoPool.cs:
```diff
@@ -0,0 +1,138 @@
+			Info = info;
+			if (Info.InitialAmmo < Info.Ammo && Info.InitialAmmo >= 0)
+				CurrentAmmo = Info.InitialAmmo;
+			else CurrentAmmo = Info.Ammo;
```
Statement on new line.

penev92 commented on OpenRA.Mods.Common/Traits/Repairable.cs:
```diff
@@ -58,10 +60,22 @@ bool CanRepairAt(Actor target)
 
+		bool CanRearmAt(Actor target)
+		{
+			return self.Info.Traits.Get<RepairableInfo>().RepairBuildings.Contains(target.Info.Name);
```
Caching this might be nice.

penev92 replied:
You are getting the Info of the trait you are in. Just cache the info in the constructor.
Same for `CanRepairAt`

penev92 commented on OpenRA.Mods.RA/Activities/LayMines.cs:
```diff
@@ -18,72 +20,97 @@
+				{
+					// Skip every AmmoPool that does not have a matching name or still has ammo
+					if (pool.Info.Name != info.AmmoPool || pool.HasAmmo())
+						continue;
```
I think this `if` is unnecessary.

reaperrr commented:
Updated & rebased.

Note for everyone else: Me and @penev92 have concluded that it's probably better to properly deal with rearm-related things like the Rearm-Repairable dependency, RearmBuildings for all actors etc. in a follow-up PR, so this doesn't grow even larger and gets delayed even more.

Updated the test-cases as well, will remove them once this PR got its +2.

reaperrr commented:
Rebased.

reaperrr commented:
Updated, fixed bug in ```HeliAttack``` activity and mammoth testcase.

penev92 commented:
I don't want to try and go through the changes for that bug. Does it have to do anything with what we discussed on IRC?

reaperrr commented:
> This is pretty inconsistent with how the airplanes to it though :( And they both do the same thing, just very differently.

Yeah, but I don't really feel like meddling with that now... as said before, I'd like to properly deal with rearm/repair in a follow-up PR.

penev92 commented:
Absolutely agreed. It was like that before this PR, it can stay a bit longer, as it has nothing to do with the original purpose of the PR.

penev92 commented on OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:
```diff
@@ -18,30 +20,35 @@ namespace OpenRA.Mods.Common.Activities
 
-			var limitedAmmo = self.TraitOrDefault<LimitedAmmo>();
-			if (limitedAmmo != null && !limitedAmmo.HasAmmo())
+			if (ammoPools != null && ammoPools.All(x => !x.Info.SelfReloads && !x.HasAmmo()))
```
Maybe add a comment above this `if` for easier readability.
`// Move to the next activity only if all ammo pools are depleted and none reload automatically

penev92 commented on OpenRA.Mods.Common/Activities/Air/HeliAttack.cs:
```diff
@@ -16,21 +19,27 @@ namespace OpenRA.Mods.Common.Activities
-			var limitedAmmo = self.TraitOrDefault<LimitedAmmo>();
-			var reloads = self.TraitOrDefault<Reloads>();
-			if (limitedAmmo != null && !limitedAmmo.HasAmmo() && reloads == null)
+			if (ammoPools != null && ammoPools.All(x => !x.Info.SelfReloads && !x.HasAmmo()))
```
Also here:
`// Return to base to reload only if all ammo pools are depleted and none reload automatically`

penev92 replied:
Btw I changed this comment intentionally. As much as I want it to be generic as the above one on `FlyAttack`, this one explicitly orders the helicopter to return to base with `HeliReturn`.
Anyway, this is what the comment should be, so hopefully we'll fix the activity soon, so it matches the one in `FlyAttack`.

penev92 commented on OpenRA.Mods.Common/Activities/Rearm.cs:
```diff
@@ -8,53 +8,64 @@
 		public override Activity Tick(Actor self)
 		{
-			if (IsCanceled || limitedAmmo == null)
+			if (IsCanceled || ammoPoolsReloadTimes == null)
```
This can be `ammoPools` instead of `ammoPoolsReloadTimes`.

penev92 replied:
Sorry, I just checked with chrisf on IRC, this should stay `ammoPoolsReloadTimes` :/

penev92 commented on OpenRA.Mods.Common/Activities/Rearm.cs:
```diff
@@ -8,53 +8,64 @@
+				needsReloading = true;
+
+			    if (--ammoPoolsReloadTimes[pool] > 0)
+			        continue;
```
Spaces detected.

penev92 commented on OpenRA.Mods.Common/Activities/Rearm.cs:
```diff
@@ -8,53 +8,64 @@
+				var hostBuilding = self.World.ActorMap.GetUnitsAt(self.Location).FirstOrDefault(a => a.HasTrait<RenderBuilding>());
 
 				if (hostBuilding == null || !hostBuilding.IsInWorld)
 					return NextActivity;
```
This really needs to go outside the `foreach`, for performance reasons, along with the `var hostBuilding = ...` and the corresponding comments

penev92 replied:
This may have been a very bad call on my part :/
The performance hit from doing this every tick instead of when we need to give 1 more ammo is probably higher than it was before you moved this :(
Which means that this hack will need to be resolved fast.
I started working on proper docking for repairing yesterday. So far it works a bit better then what's on bleed, so I should be able to do it for rearming soon (hopefully).

penev92 commented on OpenRA.Mods.Common/Activities/Rearm.cs:
```diff
@@ -8,53 +8,64 @@
+				if (!pool.GiveAmmo())
+					continue;
 
 				hostBuilding.Trait<RenderBuilding>().PlayCustomAnim(hostBuilding, "active");
```
`.PlayCustomAnim()` should be here, but move the `.Trait<RenderBuilding>()` before the `foreach`.

reaperrr replied:
It's actually better to play the anim for each ammo pool anyway.
When we refactor rearming properly in a follow-up (which includes getting rid of this ugly building hack), we can actually use this to play different animations for each ammo pool (by replacing the "active" hardcoding with something customisable), so players can get a visual indicator what exactly has been reloaded just now.

penev92 replied:
Nice idea ;)

penev92 commented on OpenRA.Mods.Common/Traits/AmmoPool.cs:
```diff
@@ -0,0 +1,138 @@
+		public readonly bool SelfReloads = false;
+
+		[Desc("Time for each ReloadCount to reload automatically.")]
+		public readonly int SelfReloadTicks = 25 * 2;
```
This description should be more consistent with the one of `ReloadTicks`. Actually, copy the fist part from there.

penev92 commented on OpenRA.Mods.Common/Traits/AmmoPool.cs:
```diff
@@ -0,0 +1,138 @@
+		[Desc("Time for each ReloadCount to reload automatically.")]
+		public readonly int SelfReloadTicks = 25 * 2;
+
+		[Desc("Whether or not reload counter should be reset when ammo has been fired.")]
```
"reload **timer**" sound better IMO.

penev92 commented on OpenRA.Mods.Common/Traits/AmmoPool.cs:
```diff
@@ -0,0 +1,138 @@
+			else
+				CurrentAmmo = Info.Ammo;
+
+			RemainingTicks = Info.ReloadTicks;
```
Shouldn't this be `Info.SelfReloadTicks` ?

reaperrr replied:
You're right, this is the 'successor' to ```Reloads.Period```. Must have gotten confused somewhere.

penev92 commented on OpenRA.Mods.RA/Activities/LayMines.cs:
```diff
@@ -18,72 +20,93 @@
+				{
+					if (pool.Info.Name == info.AmmoPool && !pool.HasAmmo())
+					{
+						// Rearm & repair at fix, then back out here to refill the minefield some more
```
No. Just **no** to "at fix".
"at repair pad" would be acceptable though.

penev92 replied:
No. Just **no** to "at fix".
"at repair pad" would be acceptable though.

penev92 replied:
No. Just **no** to "at fix".
"at repair pad" would be acceptable though.

reaperrr replied:
I'll change it to
```// Rearm (and possibly repair) at rearm building, then back out here to refill the minefield some more```

penev92 commented on OpenRA.Mods.RA/Activities/LayMines.cs:
```diff
@@ -18,72 +20,93 @@
-			if (ml.Minefield.Contains(self.Location) && ShouldLayMine(self, self.Location))
+			if (ammoPools != null)
+				foreach (var pool in ammoPools)
+				{
```
OpenRA conventions state that if this `foreach` has a block, then the outer `if` must have a block as well.

penev92 commented on OpenRA.Mods.RA/Activities/LayMines.cs:
```diff
@@ -18,72 +20,93 @@
+					if (pool.Info.Name == info.AmmoPool && !pool.HasAmmo())
+					{
+						// Rearm & repair at fix, then back out here to refill the minefield some more
+						var buildings = info.RearmBuildings;
```
Rename this to "rearmBuildings" and move it into the constructor.

penev92 replied:
Rename this to "rearmBuildings" and move it into the constructor.

penev92 commented on OpenRA.Mods.RA/Activities/LayMines.cs:
```diff
@@ -18,72 +20,93 @@
+			if (ammoPools != null)
+				foreach (var pool in ammoPools)
+				{
+					if (pool.Info.Name == info.AmmoPool && !pool.HasAmmo())
```
This `if`, `foreach`, `if` can be simplified to
`if (ammoPools != null && ammoPools.Any(pool => pool.Info.Name == info.AmmoPool && !pool.HasAmmo()))`

penev92 commented on OpenRA.Mods.RA/Activities/LayMines.cs:
```diff
@@ -18,72 +20,93 @@
 			}
 
-			// TODO: return somewhere likely to be safe (near fix) so we're not sitting out in the minefield.
+			// TODO: Return somewhere likely to be safe (near fix) so we're not sitting out in the minefield.
```
Again, please fix that "fix" in the comment.

penev92 commented on OpenRA.Mods.RA/Activities/LayMines.cs:
```diff
@@ -18,72 +20,93 @@
-			if (limitedAmmo != null)
-				limitedAmmo.TakeAmmo();
+			var info = self.Info.Traits.Get<MinelayerInfo>();
+			var ammoPools = self.TraitsImplementing<AmmoPool>();
```
I think you already have those two in the constructor.

penev92 commented on OpenRA.Mods.RA/Activities/LayMines.cs:
```diff
@@ -18,72 +20,93 @@
+			if (IsCanceled)
+				return NextActivity;
 
 			var movement = self.Trait<IMove>();
```
Move this to the constructor please.

penev92 commented on OpenRA.Mods.RA/Activities/LayMines.cs:
```diff
@@ -18,72 +20,93 @@
+			var info = self.Info.Traits.Get<MinelayerInfo>();
+			var ammoPools = self.TraitsImplementing<AmmoPool>();
+
+			if (ammoPools != null)
```
This `if`, with the `forach` and `if` it contains, can be simplified to
>			if (ammoPools != null)
			{
				var pool = ammoPools.FirstOrDefault(x => x.Info.Name == info.AmmoPool);
				if (pool != null)
					pool.TakeAmmo();
			}

penev92 commented on OpenRA.Mods.RA/Traits/Minelayer.cs:
```diff
@@ -18,19 +18,19 @@
 		[ActorReference] public readonly string[] RearmBuildings = { "fix" };
 
-		public readonly string RearmSound = "minelay1.aud";
+		public readonly string AmmoPool = "primary";
```
I think "AmmoPoolName" would be nicer.

reaperrr replied:
OK, though I'll have to make the same change for ```Armament``` then.

penev92 commented:
I think this works nicely in-game now. And I am mostly happy with how the code looks.
I left quite a few comments, but most of them are style- or performance-related. Only one was a real bug (I think).
Since his PR is really not reviewable per-commit, I would ask you to do the fixes requested above and commit them in a new "fix" commit. It would be much easier to see what changed than going through the whole thing again if you squash them. Please leave the squashing for later.

Other side notes I took while reviewing:
**1.** This could probably use a wiki page explanation. Even more so given that we will remove the test cases before merging, and we have no good test cases ingame to show off the functionality or how modders/mappers can use it.
**2.** Rearming (and namely CanRearmAt()) should really split off from Repairable one day.
**3.** This PR, among other things, should make the situation you mentioned with the 200 Yaks *a little* better. ;)

Note: As hinted above, I tested ingame and reviewed the code. I still haven't reviewed the upgrade rule and any of the YAML changes though.
But I think I can +1 this when the changes are done.

reaperrr commented:
> But I think I can +1 this when the changes are done.

Changes are done.

penev92 commented on OpenRA.Mods.Common/Activities/Rearm.cs:
```diff
@@ -8,53 +8,65 @@
+
+				needsReloading = true;
+
+			    if (--ammoPoolsReloadTimes[pool] > 0)
```
Sorry, I still see spaces here.

penev92 commented on OpenRA.Mods.Common/Activities/Rearm.cs:
```diff
@@ -8,53 +8,65 @@
 		public override Activity Tick(Actor self)
 		{
-			if (IsCanceled || limitedAmmo == null)
+			if (IsCanceled || ammoPools == null)
```
Rearming structures should only rearm ammo pools that don't self reload, right?
Or should they also speed up rearming of ammo pools that self reload?

reaperrr replied:
Good question. I guess a speedup would be rather confusing, so they should ignore ammo pools that self-reload. 
I guess that means we need to ignore self-reloading pools when populating ammoPools in the constructor, right?

penev92 replied:
Yes.

penev92 commented:
Sorry @reaperrr, looks like you'd better revert 2 of the changes from yesterday (the null check and the rearm building hack).
I left some comments on yesterday's comments for you.
I'll go ahead and check the upgrade rules and YAML changes now, and will test ingame whenever you're done.

penev92 commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -719,6 +719,65 @@ internal static void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode>
+							var limitedAmmoFields = limitedAmmo.Value.Nodes;
+							var rearmSound = minelayerFields.FirstOrDefault(rs => rs.Key == "RearmSound");
+
+							if (rearmSound != null)
```
This should not be necessary.
Even if it's `null`, you still want to add "RearmSound" to `AmmoPool`.
Also on the next line you do a proper check.

penev92 commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -719,6 +719,65 @@ internal static void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode>
+
+							if (rearmSound != null)
+							{
+								var minelayerRearmSound = rearmSound != null ? FieldLoader.GetValue<string>("RearmSound", rearmSound.Value.Value) : "";
```
This shouldn't be an empty string, but rather `"minelay1.aud"`, like you add it [here](https://github.com/OpenRA/OpenRA/pull/7324/files#diff-f696d798ec3c0ffaab7572533c827b01R448).

penev92 commented:
YAML changes look good.

reaperrr commented:
Updated.

I'll squash the fixes and remove the testcases once this has its +2.

reaperrr commented:
Rebased.

penev92 commented:
All test cases look good ingame.
The new last 4 commits look good as well, and I'm going to trust that you haven't changed the previous ones :P
I think you can squash the commits now. I don't think anyone else has reviewed this recently, so they need to start over anyway.
You can get them down to 3, although the first one could use splitting up.
Anyway, here's a long overdue :+1: 
Sorry it took so long.

reaperrr commented:
Commits squashed/fixup'd.

> You can get them down to 3, although the first one could use some splitting up.

I don't like that behemoth of a commit either, but I don't think I can split that up in a sensible way without going against the 'each commit should be compileable and working on its own' policy.

obrakmann commented on OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:
```diff
@@ -18,30 +20,36 @@ namespace OpenRA.Mods.Common.Activities
-			var limitedAmmo = self.TraitOrDefault<LimitedAmmo>();
-			if (limitedAmmo != null && !limitedAmmo.HasAmmo())
+			// Move to the next activity only if all ammo pools are depleted and none reload automatically
+			if (ammoPools != null && ammoPools.All(x => !x.Info.SelfReloads && !x.HasAmmo()))
```
I think I've said this in my previous round of reviewing, but the comment got lost.  At some point, this should really check whether there's any ammo left that is suitable for the current target. Leave a `// TODO`?

obrakmann commented on OpenRA.Mods.Common/Activities/Air/HeliAttack.cs:
```diff
@@ -41,10 +51,10 @@ public override Activity Tick(Actor self)
 
 			// Fly towards the target
-			if (!target.IsInRange(self.CenterPosition, attack.GetMaximumRange()))
+			if (!target.IsInRange(self.CenterPosition, attackHeli.GetMaximumRange()))
```
This leads to the helicopter not doing anything when the armament with the longest range is out of ammo. (As shown with the hind testcase).  `FlyAttack` uses the minimum range of all armament instead (which probably leads to planes not shooting as early as they could under circumstances). Ideally, `GetMaximumRange` and `FlyAttacks`'s `Tick()` should check for the weapon's validity against the target and the state of its ammo pool.

This behaviour isn't this PR's fault though. But maybe leave a comment?

obrakmann commented:
Code looks good, works as advertised in-game, :+1: 

Please remove the test case commit, and add those comments, or put them on your personal todo list, or open a ticket or whatever, depending on what your plans are with respect to this thing.

reaperrr commented:
Removed testcase.
Added comments AND put those issues on my todo list. Me and @penev92 intend to give rearming in general an overhaul, we'll take a closer look at these then.

penev92 commented:
[Changelog entry](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/2ab88d20ad8bc0f1ffc907cdc177ebd99a8a5d3d...7a7d69845f20e5c40c9c1266101f7f04ea2b5d52)
